packages/valory/skills/abstract_round_abci/models.py

```diff
@@ -42,27 +56,202 @@
     get_data_from_nested_dict,
     get_value_with_type,
 )
 
 
 NUMBER_OF_RETRIES: int = 5
-DEFAULT_BACKOFF_FACTOR: int = 2
+DEFAULT_BACKOFF_FACTOR: float = 2.0
 DEFAULT_TYPE_NAME: str = "str"
-_DEFAULT_REQUEST_TIMEOUT = 10.0
-_DEFAULT_REQUEST_RETRY_DELAY = 1.0
-_DEFAULT_TX_TIMEOUT = 10.0
-_DEFAULT_TX_MAX_ATTEMPTS = 10
-_DEFAULT_CLEANUP_HISTORY_DEPTH_CURRENT = None
 
 
 HeadersType = List[Tuple[str, str]]
 ParametersType = HeadersType
 
 
-class BaseParams(Model):  # pylint: disable=too-many-instance-attributes
+class FrozenMixin:  # pylint: disable=too-few-public-methods
+    """Mixin for classes to enforce read-only attributes."""
+
+    _frozen: bool = False
+
+    def __delattr__(self, *args: Any) -> None:
+        """Override __delattr__ to make object immutable."""
+        if self._frozen:
+            raise AttributeError("This object is frozen!")
+        super().__delattr__(*args)
+
+    def __setattr__(self, *args: Any) -> None:
+        """Override __setattr__ to make object immutable."""
+        if self._frozen:
+            raise AttributeError("This object is frozen!")
+        super().__setattr__(*args)
+
+
+class TypeCheckMixin:  # pylint: disable=too-few-public-methods
+    """Mixin for data classes & models to enforce attribute types on construction."""
+
+    def __post_init__(self) -> None:
+        """Check that the type of the provided attributes is correct."""
+        for attr, type_ in get_type_hints(self).items():
+            value = getattr(self, attr)
+            self.__check_type(attr, value, type_)
+
+    @classmethod
+    def __check_type(cls, name: str, value: Any, type_: Type) -> None:
+        """Check type."""
+        if type_ is Any:
+            # trivial - any type requires no check
+            return
+        if isinstance(value, MagicMock):
+            # testing - any magic value is ignored
+            return
+        generics = () if getattr(type_, "__args__", None) is None else type_.__args__
+        if len(generics) > 0 and inspect.isclass(value):
+            if not any([issubclass(value, generic) for generic in generics]):
+                raise TypeError(f"{name} must be a subclass of {generics}")
+        elif len(generics) > 0:
+            # non-trivial - many different cases possible
+            return
+        else:
+            if not isinstance(value, type_):
+                raise TypeError(f"{name} must be a {type_}, found {type(value)}")
+
+    @classmethod
+    def _ensure(cls, key: str, kwargs: Dict, type_: Any) -> Any:
+        """Get and ensure the configuration field is not None (if no default is provided) and of correct type."""
+        enforce("skill_context" in kwargs, "Only use on models!")
+        skill_id = kwargs["skill_context"].skill_id
+        enforce(
+            key in kwargs,
+            f"'{key}' of type '{type_}' required, but it is not set in `models.params.args` of `skill.yaml` of `{skill_id}`",
+        )
+        value = kwargs.pop(key)
```

Comment:
Maybe its good to change this to kwargs.get(). We would avoid situations like this: https://github.com/valory-xyz/autonomous-fund/pull/73#discussion_r1059285681
It could lead to confusing errors, i.e. complaining a key is not available when in fact it was provided.

Reply:
But the base class sets the value, why does it need to be set at all on these classes?

Reply:
Using get is not good. WE cannot warn about unset values then